File: internal/workflow/graph.go
package workflow

// graph.go 集中放纯图算法，供 validate / scheduler / show / UI 校验复用。
// 所有函数以 Definition 的 nodes（含 START/END）与 edges 为输入，不改动入参。

// BuildAdjacency 返回后继表 succ（from → []to）与前驱表 pred（to → []from），
// 均按 edges 出现顺序追加、按 node id 索引。只收录 edges 里出现的端点键，孤立节点不建键。
func BuildAdjacency(def *Definition) (succ, pred map[string][]string) {
	succ = make(map[string][]string, len(def.Nodes))
	pred = make(map[string][]string, len(def.Nodes))
	for _, edge := range def.Edges {
		succ[edge.From] = append(succ[edge.From], edge.To)
		pred[edge.To] = append(pred[edge.To], edge.From)
	}
	return succ, pred
}

// DetectCycle 用 DFS 三色标记探测环：命中返回一条环路径（如 ["a","b","a"]，首尾同 id 便于打印
// "a→b→a"）；无环返回 nil。按 nodes 顺序起 DFS，结果确定。仅依赖 edges 拓扑，与节点是否为标记节点无关。
func DetectCycle(def *Definition) []string {
	succ, _ := BuildAdjacency(def)
	const (
		white = 0 // 未访问
		gray  = 1 // 在当前 DFS 栈上
		black = 2 // 已完成
	)
	color := make(map[string]int, len(def.Nodes))
	var stack []string

	var dfs func(id string) []string
	dfs = func(id string) []string {
		color[id] = gray
		stack = append(stack, id)
		for _, next := range succ[id] {
			switch color[next] {
			case gray:
				// 命中回边：截取栈上从 next 到当前的片段，补上 next 收尾成闭环路径。
				for i, node := range stack {
					if node == next {
						cycle := append([]string{}, stack[i:]...)
						return append(cycle, next)
					}
				}
			case white:
				if cycle := dfs(next); cycle != nil {
					return cycle
				}
			}
		}
		stack = stack[:len(stack)-1]
		color[id] = black
		return nil
	}

	for _, node := range def.Nodes {
		if color[node.ID] == white {
			if cycle := dfs(node.ID); cycle != nil {
				return cycle
			}
		}
	}
	return nil
}

// Ancestors 返回从 id 沿边反向可达的全部前驱节点 id 集合（不含 id 自身）。
// 供祖先引用校验：模板 {{X}} 的 X 须是本节点祖先。可能含 START（若可达），调用方按需过滤标记节点。
// 假定图无环（有环时"祖先"无定义），故调用方应在 DetectCycle 通过后再调用。
func Ancestors(def *Definition, id string) map[string]bool {
	_, pred := BuildAdjacency(def)
	ancestors := make(map[string]bool)
	var visit func(node string)
	visit = func(node string) {
		for _, parent := range pred[node] {
			if !ancestors[parent] {
				ancestors[parent] = true
				visit(parent)
			}
		}
	}
	visit(id)
	return ancestors
}

// TopoLevels 返回 agent 节点的拓扑分层（同层可并行、不含 START / END），供 show --expand 预览。
// 层号算法：node 层号 = max(各前驱层号) + 1，START 视作 −1 层（以 START 为唯一前驱者落 level 0）；
// END 不计入。同层节点按其在 nodes[] 的出现序排列，跨次运行稳定。假定图无环（调用前应 DetectCycle 通过）。
func TopoLevels(def *Definition) [][]string {
	succ, pred := BuildAdjacency(def)

	// Kahn 拓扑序：入度 0 起，逐层松弛出层号。
	indeg := make(map[string]int, len(def.Nodes))
	for _, node := range def.Nodes {
		indeg[node.ID] = len(pred[node.ID])
	}
	level := make(map[string]int, len(def.Nodes))
	queue := make([]string, 0, len(def.Nodes))
	for _, node := range def.Nodes {
		if indeg[node.ID] == 0 {
			level[node.ID] = -1 // 源（START）视作 −1 层，其后继落 level 0
			queue = append(queue, node.ID)
		}
	}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		for _, next := range succ[current] {
			if candidate := level[current] + 1; candidate > level[next] {
				level[next] = candidate
			}
			indeg[next]--
			if indeg[next] == 0 {
				queue = append(queue, next)
			}
		}
	}

	// 按层号聚合 agent 节点，同层保持 nodes[] 出现序。
	maxLevel := -1
	for _, node := range def.Nodes {
		if node.IsAgent() && level[node.ID] > maxLevel {
			maxLevel = level[node.ID]
		}
	}
	if maxLevel < 0 {
		return nil
	}
	levels := make([][]string, maxLevel+1)
	for _, node := range def.Nodes {
		if node.IsAgent() {
			lvl := level[node.ID]
			// 无入边的 agent 落在 −1 层（合法图里只有 START 无入边，agent 无入边即非法输入）：
			// 跳过而非 levels[-1] 越界 panic。TopoLevels 约定输入已过校验，此处仅作防御。
			if lvl < 0 {
				continue
			}
			levels[lvl] = append(levels[lvl], node.ID)
		}
	}
	return levels
}

// AgentNodeIDs 返回 agent 节点 id 流（排除 START / END），按确定性拓扑序（先层号、同层再按 nodes[] 出现序）。
// 异常图（拓扑分层未覆盖全部 agent 节点，如手改成环）按 nodes[] 出现序兜底补上，不丢节点。CLI `workflow list`
// 与 UI 运行详情共用此单一实现——两端节点流保证一致、不漂移。空定义返回空切片（JSON 里为 []、不为 null）。
func AgentNodeIDs(def *Definition) []string {
	ids := make([]string, 0, len(def.Nodes))
	seen := make(map[string]bool, len(def.Nodes))
	for _, level := range TopoLevels(def) {
		for _, id := range level {
			ids = append(ids, id)
			seen[id] = true
		}
	}
	for _, node := range def.Nodes {
		if node.IsAgent() && !seen[node.ID] {
			ids = append(ids, node.ID)
		}
	}
	return ids
}

File: internal/ui/assets/js/graph.js
// DAG 图算法：镜像 internal/workflow/graph.go，供编辑器做即时校验 / 拖拽连边环检测 / 占位符祖先
// 建议 / 整理连线（传递归约）。运行详情页的进度画布另有自己的分层布局，见 dag-layout.js。
//
// 输入统一为 { nodes: [{id, ...}], edges: [{from, to}] } 形状（含 START/END），纯函数、不改入参。
// 前端这份镜像须与 Go 版规则保持同步（见 ui.md〈已知限制〉：前端镜像校验可能与内核漂移）。

export const NODE_ID_START = "START";
export const NODE_ID_END = "END";

export function isMarker(id) {
  return id === NODE_ID_START || id === NODE_ID_END;
}

export function isAgent(id) {
  return !isMarker(id);
}

// buildAdjacency 返回后继表 succ / 前驱表 pred（Map<id, id[]>），只收录 edges 里出现的端点键。
export function buildAdjacency(def) {
  const succ = new Map();
  const pred = new Map();
  for (const edge of def.edges || []) {
    pushInto(succ, edge.from, edge.to);
    pushInto(pred, edge.to, edge.from);
  }
  return { succ, pred };
}

function pushInto(map, key, value) {
  if (!map.has(key)) map.set(key, []);
  map.get(key).push(value);
}

// detectCycle：DFS 三色标记探测环，命中返回一条环路径（如 ["a","b","a"]，首尾同 id 便于打印
// "a→b→a"）；无环返回 null。按 nodes 顺序起 DFS，结果确定。
export function detectCycle(def) {
  const { succ } = buildAdjacency(def);
  const WHITE = 0,
    GRAY = 1,
    BLACK = 2;
  const color = new Map();
  const stack = [];

  function dfs(id) {
    color.set(id, GRAY);
    stack.push(id);
    for (const next of succ.get(id) || []) {
      const c = color.get(next) || WHITE;
      if (c === GRAY) {
        const i = stack.indexOf(next);
        return [...stack.slice(i), next];
      }
      if (c === WHITE) {
        const cycle = dfs(next);
        if (cycle) return cycle;
      }
    }
    stack.pop();
    color.set(id, BLACK);
    return null;
  }

  for (const node of def.nodes || []) {
    if ((color.get(node.id) || WHITE) === WHITE) {
      const cycle = dfs(node.id);
      if (cycle) return cycle;
    }
  }
  return null;
}

// wouldCreateCycle：模拟加入 edge 后是否成环（供拖拽连边即时判断），不修改 def。
export function wouldCreateCycle(def, edge) {
  const probe = { nodes: def.nodes, edges: [...(def.edges || []), edge] };
  return detectCycle(probe) !== null;
}

// ancestors：沿边反向可达的全部前驱 id 集合（不含自身）。与 Go 版不同：本实现自带 visited 防御，
// 图暂时有环（编辑中间态、尚未通过校验）也不会死循环——Go 版假定调用前已过 DetectCycle，但编辑器
// 需要在校验通过之前也能安全渲染画布 / 算占位符建议，故加此防御（前端比 Go 版更宽容，是刻意的
// UI 侧加固，不影响合法图的结果）。
export function ancestors(def, id) {
  const { pred } = buildAdjacency(def);
  const out = new Set();
  const visiting = new Set();
  const visit = (node) => {
    if (visiting.has(node)) return; // 防环：已在当前访问链上，不再下钻
    visiting.add(node);
    for (const parent of pred.get(node) || []) {
      if (!out.has(parent)) {
        out.add(parent);
        visit(parent);
      }
    }
  };
  visit(id);
  return out;
}

// edgeKey：边的去重键（node id 不含 \u0000，故安全）。供 dag-layout / run-detail / editor 复用。
export function edgeKey(edge) {
  return edge.from + "\u0000" + edge.to;
}

// redundantEdges：DAG 传递归约——找出全部"冗余边"。边 from→to 冗余，当且仅当 from 另有后继 w（w≠to）
// 能到达 to，即存在 from→w→…→to 的绕行路径；此时直连 from→to 不改变任何可达性，可安全删除
// （例：START→node2 因 START→node1→node2 而多余）。返回冗余边数组 [{from, to}]。传递归约保持传递
// 闭包不变，故不会破坏单源单汇 / 节点度约束，且多条冗余边可同时删除。要求无环（有环时传递归约无
// 定义）；reach 记忆化时先占位空集，编辑中间态即便暂时有环也只拿到部分结果而非无限递归——结果仅
// 在无环图上具传递归约语义，此防御只保证不死循环。
export function redundantEdges(def) {
  const { succ } = buildAdjacency(def);
  // reach(x)：x 沿边可达的节点集合（≥1 步，不含 x 自身）。
  const memo = new Map();
  const reach = (x) => {
    const cached = memo.get(x);
    if (cached) return cached;
    const out = new Set();
    memo.set(x, out); // 先登记空集占位：有环时回引拿到部分结果而非无限递归
    for (const next of succ.get(x) || []) {
      out.add(next);
      for (const far of reach(next)) out.add(far);
    }
    return out;
  };
  const redundant = [];
  for (const e of def.edges || []) {
    const outs = succ.get(e.from) || [];
    // 存在另一后继 w 能到达 e.to → e.from→e.to 有绕行路径，冗余。
    if (outs.some((w) => w !== e.to && reach(w).has(e.to))) redundant.push({ from: e.from, to: e.to });
  }
  return redundant;
}
